src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
+
+    })(Strategy.Sequential)
+
+    tryCompleteTask(task.get, actor)
+    () => {actor ! Interrupt}
+  }
+
+
+  /**
+   * Asynchronous execution of this Process. Note that this method is not resource safe unless
+   * callback is called with _left_ side completed. In that case it is guaranteed that all cleanups
+   * has been successfully completed.
+   *
+   * User of this method is responsible for any cleanup actions to be performed by running the
+   * next Process obtained on right side of callback.
+   *
+   * This method returns a function, that when applied, causes the running computation to be interrupted.
+   * That is useful of process contains any asynchronous code, that may be left with incomplete callbacks.
+   * If the evaluation of the process is interrupted, then the interruption is only active if the callback
+   * was not completed before, otherwise interruption is no-op.
+   *
+   * This implementation also guarantees that when `cb` is invoked on right side, then sequence is always non-empty.
+   * This is usefully to run two processes truly in parallel w/o consulting higher level combinator (i.e. wye/merge).
+   *
+   * This implementation as well guaranteed that any onComplete // onFail // onKill code is consulted exactly once
+   * whether the computation was or was not interrupted.
```

Comment:
> The problem will be in this type of programs…

Given how pervasive this idiom is with resource management in scalaz-stream, I think we very definitely need to get it right!  You yourself said that your code is very dependent on `onComplete` getting evaluated; I know for a fact that our code would start leaking resources if we were to ignore this use-case.

My opinion is that if the awaited task completes *despite* our interrupt, then we absolutely need to run any finalizers which are dependent on its value.  Conveniently enough, these finalizers will be contained exclusively within the body of the await, and even more conveniently, those finalizers will *not* have been run previously by the interrupt!  Thus, it is very easy to implement this semantic.  The only weird thing is that the `onComplete` of the outer process (what I call the "outer finalizer" in my test case) will be run before the `onComplete` of the inner process (what I call the "inner finalizer"), but I don't think that's the worst thing in the world.

> then we can introduce this: [`awaitInterruptibly`]

This has a very similar signature to `awaitOr`, which can in fact fulfill exactly the same use-case so long as we parameterize `EarlyCause` with `A` and add an `Interrupted[A]` case to the algebra.